text/0003-placeholders-in-conditions.md

```diff
@@ -0,0 +1,157 @@
+
+2. Template-based policy construction prevents code injection attacks (i.e., in the style of SQL injection). Templates act like prepared statements, where placeholders can only be instantiated with data values; they cannot be instantiated with strings that could be reinterpreted as arbitrary policy code. Using templates whenever possible is an important protection for applications that add policies as a result of user actions.
+
+This RFC was motivated by a particular customer use-case, i.e., [#81](https://github.com/cedar-policy/cedar/issues/81).
+
+# Detailed design
+
+This RFC does not propose any change to what syntax is valid in the policy scope. For instance, the following remains _invalid_:
+```
+permit(
+  principal,
+  action,
+  ?resource in resource
+);
+```
+The only valid uses of `?resource` in the policy scope are `resource == ?resource` and `resource in ?resource`, and this RFC is not proposing to change that. This RFC also disallows the use of `?principal` and `?resource` in the condition if they don't _also_ appear in the scope.
+
+Both restrictions are in place because these two placeholder variables have special meaning for indexing policies in a policy store. In particular, looking _only_ at the linking information (i.e., to what `?principal` and/or `?resource` are bound) is sufficient to precisely select linked policies for particular requests. If the variables could only be bound in the condition or in other positions in the scope, and used in arbitrary contexts, it would lead to selecting superfluous policies.
+
+## Fresh placeholder variables
+
+To support a policy like the one above, this RFC proposes support for placeholder variables beyond `?resource` and `?principal`. With their support, we can write:
+```
+template (?lowerbound)
+permit(principal, action, resource)
+when { ?lowerbound in resource };
+```
+Here, the `template (?lowerbound)` part _introduces_ a fresh placeholder variable which can be used legally within the policy template that follows. Eliding the `template (?lowerbound)` prefix would be illegal, signaling an unbound variable.
+
+Multiple variables can be introduced together, separated by commas, e.g.,
+```
+template (?user, ?resourcebound)
+permit(principal, action, resource)
+when { 
+     ?resourcebound in resource 
+  && principal in ?user.delegates
+};
+```
+
+Policy writers can choose to include `?principal` and/or `?resource` in the list of template variables if they like, but doing so is not required. These two variables retain their special status.
+
+If a variable is included in the list that does not appear in the actual template, we can issue a warning.
+
+While all introduced placeholder variables presented so far are assumed to be entities (just like `?principal` and `?resource`), there is no reason why they cannot have arbitrary type. The particular type of the placeholder need not be given, for simplicity.
```

Comment:
"there is no reason" are arbitrary-typed placeholder variables in scope for this RFC, or is that a proposed future extension? If this will be included as part of the current RFC it would be nice to see some concrete examples. The string and long cases are easy, but I wonder if there's anything difficult about supporting placeholder variables that have type record or set.